iothub/device/tests/Pipeline/ConnectionStateDelegatingHandlerTests.cs

```diff
@@ -0,0 +1,336 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Microsoft.Azure.Devices.Client.Transport;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Moq;
+
+namespace Microsoft.Azure.Devices.Client.Tests
+{
+    [TestClass]
+    [TestCategory("Unit")]
+    public class ConnectionStateDelegatingHandlerTests
+    {
+        [TestMethod]
+        public async Task ConnectionStateDelegatingHandler_CloseAsyncBeforeOpen_Ok()
+        {
+            // arrange
+            var nextHandlerMock = new Mock<IDelegatingHandler>();
+            var contextMock = new PipelineContext
+            {
+                RetryPolicy = new IotHubClientTestRetryPolicy(1),
+                ConnectionStatusChangeHandler = (connectionStatusInfo) => { }
+            };
+            using var sut = new ConnectionStateDelegatingHandler(contextMock, nextHandlerMock.Object);
+
+            // act and assert
+            await sut.CloseAsync(CancellationToken.None).ConfigureAwait(false);
+        }
+
+        [TestMethod]
+        public async Task ConnectionStateDelegatingHandler_OpenAsync_SynchronizesToOneInnerCallToOpen()
+        {
+            // arrange
+            var contextMock = new PipelineContext
+            {
+                RetryPolicy = new IotHubClientNoRetry(),
+                ConnectionStatusChangeHandler = (connectionStatusInfo) => { }
+            };
+
+            int callCounter = 0;
+            var nextHandlerMock = new Mock<IDelegatingHandler>();
+            nextHandlerMock
+                .Setup(x => x.OpenAsync(It.IsAny<CancellationToken>()))
+                .Returns(() =>
+                {
+                    callCounter++;
+                    return Task.CompletedTask;
+                });
+            // Artificial wait to simulate the transport protocol level async operations.
+            // Without this, the transport layer will immediately transition to Closed, which will cause the "Open" flow to be executed again.
+            // The internal thread pool will determine how the parallel OpenAsync() calls in act are scheduled.
+            nextHandlerMock
+                .Setup(x => x.WaitForTransportClosedAsync())
```

Comment:
Review and remove wherever not required

Reply:
Is this the same as openAsync on already opened device client? Or that would be a separate test?

Reply:
This test is for specifically testing that parallel openasync calls are synchronized in the connection state management class. I'm not sure if we have a test for calling open on an already open client, but I can add one.